### ISSUE TITLE
Correct ordering of DICOMs during conversion

### ISSUE THREAD

**Issue by jan-petr** [ExploreASL/ExploreASL#514]
### Idea

- Add %r to `io_dcm2nii` in line 132 👍 
- dcm2nii_args = '-f "%f_%p_%t_%s_%r"'; 👍 
- Extract the instance number. 👍 
- Sort renamed ASL by instance number 👍 
- return initial instance numbers for all converted NIFTIs 👍 

xASL_imp_DCM2NII_Subject
- Remove functions xASL_imp_DCM2NII_Subject_RealASLContext and xASL_imp_DCM2NII_Subject_SwitchNIIs 👍 

xASL_imp_DCM2NII_Subject_StoreJSON
- Give paths and instance numbers to all creates JSONs
- Make sure that the extracted information is saved to JSON based on instance numbers
- Create the outputs separately for each JSON - change parms to a cell - each cell for each JSON, write information to parms based on the range of instanceNumber.
- We have to do it in one run, so that we read DICOMs only once.
- This completes the JSON information for each separate NIFTI

xASL_imp_DCM2NII_Subject_ShuffleTheDynamics
- We move this and run it only after xASL_imp_DCM2NII_Subject_StoreJSON, that way, we have more information for the merging of files.
- We have to make sure that the JSON fields are correctly as this was all done previously in StoreJSON.

**Comment by MichaelStritt** [comment]
Just found a minor bug related to my changes within xASL_bids_Dicom2JSON where I was working on the ImageTypes/InstanceNumbers array. I should remove that part from there if we change this anyway.

```
Undefined function or variable 't_parms'.
Error in xASL_bids_Dicom2JSON (line 324)
                    if isfield(t_parms,fieldname)
Error in xASL_imp_DCM2NII_Subject>xASL_imp_DCM2NII_Subject_StoreJSON (line 294)
                [parms, pathDcmDict] = xASL_bids_Dicom2JSON(imPar, Fpath, SavePathJSON{iPath}, imPar.dcmExtFilter, bUseDCMTK, pathDcmDict);
```

**Comment by MichaelStritt** [comment]
Hmm... I'm trying to find the Philips bug, but it's not that obvious. Within `xASL_bids_Dicom2JSON` there is this part:

```matlab
%% -----------------------------------------------------------------------------
% Obtain the selected DICOM parameters from the header
% Write the new parameter to the list (or put the default value)
% -----------------------------------------------------------------------------
for iField=1:length(dcmfields)
                fieldname = dcmfields{iField};
                
                % Extract value of current field
                if  isfield(temp, fieldname)
                    thevalue = temp.(fieldname);
                    thevalue = thevalue(~isnan(thevalue));
                else
                    thevalue = [];
                end
                
                % Convert string-numbers to numbers if necessary
                if ~isempty(thevalue)
                    if length(xASL_str2num(thevalue))>1
                        tmpTheValue = nonzeros(thevalue);
                        tmpTheValue = tmpTheValue(1);
                    else
                        tmpTheValue = thevalue;
                    end
                    t_parms{parmsIndex}.(fieldname)(iMrFile) = xASL_str2num(tmpTheValue);
                else
                    if imPar.bVerbose
                        if iMrFile==1, fprintf('%s\n',['Parameter ' fieldname ' not found, default used']); end
                    end
                    t_parms{parmsIndex}.(fieldname)(iMrFile) = DcmParDefaults.(fieldname);
                end
end
```

If I understood everything correctly, then we check the `fieldname`s here. One of the problematic fields is `RWVSlope`, because that's returned as `NaN`, which leads to some errors within the `xASL_adm_GetPhilipsScaling` function, right?
The field is already `NaN` within the `temp` struct, so we have to look at an earlier part, but `temp` comes directly out of `xASL_io_DcmtkRead`, so there's nothing we can do here, right? So we just have to set the default value for `RWVSlope` correctly? Or am I on the completely wrong track?
I also saw that the NIfTIs coming out of a subfunction of `xASL_io_dcm2nii` have a weird name format. The first flavor has two NIfTIs looking like this `ASL__20150709175853_501_1_real` for example. In the end the name seems to be fine though.

**Comment by jan-petr** [comment]
I'll take it from here - that's a minor thing that I broke - I think that I know where to fix this.

**Comment by MichaelStritt** [comment]
@jan-petr: In the commits 0a93075 to f597162 that are shown above I restructured some of the variables to struct fields of x. We wanted to start with this regarding #480 anyway. It also helps with code readability, since the `_DCM2NII_Subject` & `_DCM2NII_CreateSummaryFile` scripts don't have a million input and output parameters anymore.

### Minor tests to show that everything still works

I ran all GE datasets and encountered these bugs, but they are still related to the ordering issues and the os related differences, right?

```
Dataset: GE_PCASL_3Dspiral_Product_volunteer
File:	\sub-Sub1\perf\sub-Sub1_asl.json
	Different dimension: LabelingDuration
	Different dimension: PostLabelingDelay
File:	\sub-Sub1\perf\sub-Sub1_asl.nii.gz
	RMSE     (1.679016e+00)
	MIN-DIFF (0)
	MAX-DIFF (0)
Dataset: GE_PCASL_3Dspiral_volunteer
File:	\sub-Sub103\perf\sub-Sub103_asl.nii.gz
	RMSE     (1.693182e+00)
	MIN-DIFF (0)
	MAX-DIFF (0)
```

**Comment by MichaelStritt** [comment]
@jan-petr: In 6a29939 I revamped `NII2BIDS_Subject`. Now, most of the scripts have less than 500 lines of codes. I tested with all GE flavors again and got the same results as above. To make sure there are no resulting bugs in the pipeline, I'm processing the GE flavors right now.

**Comment by jan-petr** [comment]
> @jan-petr: In the commits [0a93075](https://github.com/ExploreASL/ExploreASL/commit/0a93075aab38064b5e14db0b7d7340e9df7b436b) to [f597162](https://github.com/ExploreASL/ExploreASL/commit/f5971629af26db2df001ce9577b4e29aa9248492) that are shown above I restructured some of the variables to struct fields of x. We wanted to start with this regarding #480 anyway. It also helps with code readability, since the `_DCM2NII_Subject` & `_DCM2NII_CreateSummaryFile` scripts don't have a million input and output parameters anymore.
> 
> ### Minor tests to show that everything still works
> I ran all GE datasets and encountered these bugs, but they are still related to the ordering issues and the os related differences, right?
> 
> ```
> Dataset: GE_PCASL_3Dspiral_Product_volunteer
> File:	\sub-Sub1\perf\sub-Sub1_asl.json
> 	Different dimension: LabelingDuration
> 	Different dimension: PostLabelingDelay
> File:	\sub-Sub1\perf\sub-Sub1_asl.nii.gz
> 	RMSE     (1.679016e+00)
> 	MIN-DIFF (0)
> 	MAX-DIFF (0)
> Dataset: GE_PCASL_3Dspiral_volunteer
> File:	\sub-Sub103\perf\sub-Sub103_asl.nii.gz
> 	RMSE     (1.693182e+00)
> 	MIN-DIFF (0)
> 	MAX-DIFF (0)
> ```

Weird. After my last revamp, I had a different order than one week ago and I have updated the order in the Flavors reference. Have you pulled the new Flavors?

**Comment by MichaelStritt** [comment]
I think I'm sadly up to date :/
```
m.stritt@MEDIRI09-NB MINGW64 /m.stritt/Server_xASL/FlavorDatabase (main)
$ git log
commit 778d05c6fcd0049ffaffaa3563d6e82613ab105f (HEAD -> main, origin/main, origin/HEAD)
Author: Petr <j.petr@hzdr.de>
Date:   Wed Apr 28 13:29:04 2021 +0200

    Improved reference fields in Philips

```

**Comment by MichaelStritt** [comment]
The rest seems to be fine, at least for GE [logContent.xlsx](https://github.com/ExploreASL/ExploreASL/files/6411304/logContent.xlsx)

**Comment by MichaelStritt** [comment]
@jan-petr: But you only changed the reference JSONs, right? I quickly checked your last commits and I can't see any changed NIfTIs. If the reference NIfTI is in the wrong order, then the comparison has to show these differences.

**Comment by MichaelStritt** [comment]
Okay, so from my side, the reference of `GE_PCASL_3Dspiral_Product_volunteer` has `m0scan,deltam` and mine has `deltam,m0scan`.  The DICOMs seem to have `deltam,m0scan` as well, so I think the reference is still wrong, right?

**Comment by jan-petr** [comment]
Sorry - I have committed and not pushed :D Try again :) The reference should have deltam,m0scan now correctly for both Linux/Windows/Reference.

**Comment by jan-petr** [comment]
Now all GE works for me and most of Philips.

Philips_PCASL_2DEPI_Bsup_AD1
Philips_PCASL_2DEPI_Bsup_AD3
Philips_PCASL_2DEPI_glioma
Philips_PCASL_3DGRASE_R5.4_TopUp

Still hang on some minor issues - will fix that on Monday,

**Comment by jan-petr** [comment]
A small bug fixed, now down to these that do not match the reference.
Philips_PCASL_2DEPI_Bsup_AD3
Philips_PCASL_2DEPI_glioma
Philips_PCASL_3DGRASE_R5.4_TopUp
Siemens_PCASL_2DEPI_AD
Siemens_PCASL_2DEPI_pharma

**Comment by jan-petr** [comment]
There was a bug in EffectiveEchoSpacing in the previous reference. This was now corrected in the current code and Reference updated. Remaining errors:

Philips_PCASL_3DGRASE_R5.4_TopUp
Siemens_PCASL_2DEPI_AD
Siemens_PCASL_2DEPI_pharma

**Comment by jan-petr** [comment]
The last one is now fine. 

These two are the only remaining problematic. There seems to be something wrong when renaming files freshly after the dcm2nii. 
I will check that on Monday - we have to change that carefully since it is working perfectly for all the remaining datasets.

Philips_PCASL_3DGRASE_R5.4_TopUp
Siemens_PCASL_2DEPI_AD

**Comment by MichaelStritt** [comment]
### Siemens_PCASL_2DEPI_AD

In `xASL_io_dcm2nii.m` lines 340 and following, we do the extraction of the instance number from the scan name. I saw some weird stuff there before, when I checked out the FME sequence. There dcm2nii just added an `e` and another number to the filename. 

Here we get:

```
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Convert 46 DICOM as .\dcm2nii_temp_ASL4D\ASL__20170914145716_24_1 (64x64x21x46)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Convert 46 DICOM as .\dcm2nii_temp_ASL4D\ASL__20170914145716_30_1 (64x64x21x46)
Conversion required 0.881000 seconds.
```

I assume `24` and `30` are the instance numbers? So it adds an unexpected `_1` ?
Why do we even get two separate NIfTIs here? There is only an ASL scan and `M0=false`. So from the 46 ASL DICOMs we should only get a single `ASL` in the first place? Or is there a difference for MOSAIC data?

**Comment by MichaelStritt** [comment]
### Philips_PCASL_3DGRASE_R5.4_TopUp

Just tested this one as well. I really don't understand the format of the M0 files though. Are those DICOMs without a file extension?

```
xASL_Move: skip: destination exists: .\TmpConversion\Philips_PCASL_3DGRASE_R5.4_TopUp\analysis\Sub1\ASL_1\M0_1.nii
xASL_Move: skip: destination exists: .\TmpConversion\Philips_PCASL_3DGRASE_R5.4_TopUp\analysis\Sub1\ASL_1\M0_1.json
```

The deltam scans have series number `801`,  the M0 scans starting with AP have series number `601` and then there are also these scans within the M0 folder that start with the filename RL and have series number `701`.
The main problem is that at a certain point we do get two resulting M0 NIfTIs, at least I think that's the problem. Makes sense, because the M0 scans have a differing series number. But we our script isn't prepared to merge that and I'm not even sure if we should? Do you know what those RL files are?

**Comment by jan-petr** [comment]
> ### Siemens_PCASL_2DEPI_AD
> In `xASL_io_dcm2nii.m` lines 340 and following, we do the extraction of the instance number from the scan name. I saw some weird stuff there before, when I checked out the FME sequence. There dcm2nii just added an `e` and another number to the filename.
> 
> Here we get:
> 
> ```
> Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
> Convert 46 DICOM as .\dcm2nii_temp_ASL4D\ASL__20170914145716_24_1 (64x64x21x46)
> Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
> Convert 46 DICOM as .\dcm2nii_temp_ASL4D\ASL__20170914145716_30_1 (64x64x21x46)
> Conversion required 0.881000 seconds.
> ```
> 
> I assume `24` and `30` are the instance numbers? So it adds an unexpected `_1` ?
> Why do we even get two separate NIfTIs here? There is only an ASL scan and `M0=false`, so maybe `1` is the correct instance number? So from the 46 ASL DICOMs we should only get a single `ASL` in the first place? Or is there a difference for MOSAIC data?
> 
> Edit: Sorry, I get it now. So `24` and `30` are the correct series numbers and `1` is the correct instance number for each scan. But here we have normal ASL scans (series number `24`) and these weird `Sname` files (series number `30`).

Yes - Siemens, and that's not the most complicated example. The good thing was that this was previously working without problems, so we just have to adapt it that it works with our new InstanceNumber-scheme.

Series Numbers and InstanceNumbers 24_1 and 30_1 sound plausible. I don't see the 'e' and 'sname' in your comment in the filename, but my guess is that this is something we have obtained previously and it signifies difference in controls and labels.

So lets just see if I manage to correctly import it and I'll check if we can extract control/label position and other features later... Not a huge problem IMHO.

**Comment by jan-petr** [comment]
> ### Philips_PCASL_3DGRASE_R5.4_TopUp
> Just tested this one as well. I really don't understand the format of the M0 files though. Are those DICOMs without a file extension?
> 
> ```
> xASL_Move: skip: destination exists: .\Philips_PCASL_3DGRASE_R5.4_TopUp\analysis\Sub1\ASL_1\M0_1.nii
> xASL_Move: skip: destination exists: .\Philips_PCASL_3DGRASE_R5.4_TopUp\analysis\Sub1\ASL_1\M0_1.json
> ```
> 
> The deltam scans have series number `801`, the M0 scans starting with AP have series number `601` and then there are also these scans within the M0 folder that start with the filename RL and have series number `701`.
> The main problem is that at a certain point we do get two resulting M0 NIfTIs, at least I think that's the problem. Makes sense, because the M0 scans have a differing series number. But our script isn't prepared to merge that and I'm not even sure if we should? Do you know what those RL files are?

Yes, DICOMs do not need to have a file-extension. AP and RL means phase-encoding in the Anterior-posterior or right-left direction - that's useful if you want to estimate the susceptibility deformations. But we can of course throw the RL away. Will check - thanks.

**Comment by jan-petr** [comment]
Partly fixed, we now consider also the SeriesNumber, and all works. Except two datasets - I have actually noticed that these two datasets were not working previously either, because they were also producing two M0 files and we just renamed them to M0.nii and M0PERev.nii which was then later on treated in the BIDS script as reversed PE direction and saved correctly.

Since we now add SeriesNumber and InstanceNumber to the file after conversion, we have to consider this in the manual renaming - I have fixed that as well, so it should all run. But I need to create a new issue as this has to be solved in a better way (and correctly for the Philips dataset which is not AP-PA but AP-RL! New issue for that is #534 - I will fix that in 1.7.0

Remaining issues (for @MichaelStritt):
1. Siemens_PCASL_2DEPI_AD - We are again facing this issue with incorrect rounding while writing SliceTiming (not that critical, but would be nice to fix or create an issue for 1.7.0).
2. When comparing BIDS folders, it does not report files/directories that are in Reference but missing in Source. I.e. if the BIDS conversion completely fails and doesn't write anything, it is not reported on screen. I just hope that once we fix this, we don't discover 20 new errors in the conversion :D #535

Pull new code and new Flavors and all should be working now...
